### PR TITLE
remove Visitor from checkSharedAccess()

### PR DESCRIPTION
**Pull request by WalterBright**
replace with nested functions

### CODE REVIEW

**Comment by maxhaton** [comment]
What is the rationale for changing it away from a visitor (I think in this case it might actually help with https://issues.dlang.org/show_bug.cgi?id=22626 because it currently skips over cast expressions, but still)?

**Comment by WalterBright** [comment]
I just noticed that the original seems to check the entire expression AST, over and over and over, when all it needs to do is check the top level (like checkAccess() does). It's one of the reasons I don't like the Visitor pattern, it can be really hard to see what is going on.

Another reason I don't like Visitor is it makes it impossible to add `const` to the checks.

File: src/dmd/expressionsem.d [hunk 1]
```diff
@@ -12805,129 +12805,127 @@ bool checkSharedAccess(Expression e, Scope* sc, bool returnRef = false)
         return false;
     }
 
-    //printf("checkSharedAccess() %s\n", e.toChars());
+//xyzzy
```
Comment (dkorpel) [inline src/dmd/expressionsem.d:12808]: debugging remnant

**Review by dkorpel — APPROVED** [review]

File: src/dmd/expressionsem.d (hunk 1 above)
Comment (WalterBright) [reply src/dmd/expressionsem.d]: removed

**Review by ljmf00 — APPROVED** [review]

**Review by RazvanN7 — APPROVED** [review]

**Pull request merged**